Full firmware loader + support for 0x08 records [ROM patch]

CTCaer opened the pull request:
The loader can now create known ROM and RAM regions and load dumps provided.
Lastly, it now asks if user wants to patch the ROM+RAM regions.

shuffle2 commented on brcm_fwpatch_ldr.py:
```diff
@@ -14,15 +14,25 @@ def accept_file(li, n):
-    # use bss to tell what is not patched
-    idaapi.add_segm_ex(seg, 'patchram', 'BSS', idaapi.ADDSEG_NOSREG)
+def make_seg(addr_range, type):
+    if type:
```
Looks like duplicate code here and type is a python keyword. Could you change it to another name and remove duplication?

CTCaer replied:
I'll do my best

shuffle2 reviewed (changes requested):
Thanks! Minor comments..

CTCaer commented:
Ok, I removed deduplication and keyword. It can also accept generic types like before.

shuffle2 commented on brcm_fwpatch_ldr.py:
```diff
@@ -14,15 +14,18 @@ def accept_file(li, n):
+        seg.type = idaapi.SEG_CODE
+        # use bss to tell what is not patched
+        idaapi.add_segm_ex(seg, 'patchram', 'BSS', idaapi.ADDSEG_NOSREG)
+        if record_type == 0x08:
```
looks like it can just be:
```python
seg.perm = idaapi.SEGPERM_READ | idaapi.SEGPERM_EXEC
if record_type != 0x08:
  seg.perm |= idaapi.SEGPERM_WRITE
idaapi.add_segm_ex(seg, 'patchram', 'BSS', idaapi.ADDSEG_NOSREG)
```
btw, what is type 8 used for?

edit: if type 8 is used to patch rom, probably want to change things a bit. Perhaps either load the rom image to 0 or (if file not found?) or create an empty code segment in that space. Then the patches will work better in ida (and each type 8 patch does not need to create a segment).

CTCaer replied:
Hmm.

What if it does the following:
rom1:
Create seg with addr: x0 size: xC8000
Asks to load additional binary and then use patch_many_bytes with 0x08 records,
else empty bytes

rom2:
Create seg with addr: 0x260000 size: xC000
Asks to load additional binary,
else empty bytes

ram_lo:
Create seg with addr: 0x0D0000 size: x10000
ram_hi:
Create seg with addr: 0x200000 size: x48000

And then use patch_many_bytes with 0x0A records to patch ram_lo and ram_hi.

So basically it will create the basic segments and then user can identify the patchram data by patched bytes.

-----------

BTW, the above, will work correctly when we get the unpatched ROM, by using a patchram with 0x08 stripped (except if it can boot with no patchram). 
I would do that, but I don't know if the device will boot in a working state.
And I don't have the spi programmer yet, so I don't want to lose a 2nd Joy-Con. 

Lastly, it seems that sub_6EC0E handles patchram.

shuffle2 replied:
That sounds good, now that memory layout is known :)

CTCaer commented:
I went ahead and done the changes.

The commit is in working state, but it's mainly to get feedback.

When everything will be OK, I'll squash the commits for the loader.

shuffle2 commented:
Seems OK :)

CTCaer committed:
Implement 0x08 record [ROM patch]

CTCaer committed:
Update loader w/ full firmware support

The loader now does the following:
*Creates ROM and RAM regions.
*Asks to load dumps in those regions.
*Ask to patch them with the PatchRAM image provided.

Supports 0x08 record [ROM patch]

CTCaer commented:
Ready for review and merge
